### PR TITLE
Zephyr add memory mapping for R-car gpio driver 

### PR DESCRIPTION
**Pull request by xakep-amatop**
Add memory mapping to driver in order to use driver with platforms based on Cortex A with enabled MMU.

### CODE REVIEW

File: drivers/gpio/gpio_rcar.c [hunk 1]
```diff
@@ -48,51 +46,50 @@ struct gpio_rcar_data {
 #define OUTDTSEL 0x40   /* Output Data Select Register */
 #define BOTHEDGE 0x4c   /* One Edge/Both Edge Select Register */
 
-static inline uint32_t gpio_rcar_read(const struct gpio_rcar_cfg *config,
+static inline uint32_t gpio_rcar_read(const struct device *dev,
 				      uint32_t offs)
 {
-	return sys_read32(config->reg_addr + offs);
+	return sys_read32(DEVICE_MMIO_GET(dev) + offs);
```
Comment (henrikbrixandersen) [inline drivers/gpio/gpio_rcar.c:52]: Please use `DEVICE_MMIO_NAMED_GET()` instead. You might find some inspiration in `drivers/gpio/gpio_intel.c`.

File: include/zephyr/drivers/gpio.h [hunk 2]
```diff
@@ -456,6 +457,7 @@ struct gpio_driver_config {
  * element in the driver's struct driver_data declaration.
  */
 struct gpio_driver_data {
+	DEVICE_MMIO_RAM; /* Must be first */
```
Comment (henrikbrixandersen) [inline include/zephyr/drivers/gpio.h:460]: NAK on these changes as they will increase the ROM/RAM footprint needlessly for all other GPIO drivers. Please seek inspiration in `drivers/gpio/gpio_intel.c` on how to contain MMIO within the GPIO driver itself.

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]
Please split this PR into two: One containing the changes to the R-Car GPIO driver, one for adding regulator-gpio support.

File: drivers/regulator/regulator_gpio.c [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+struct regulator_gpio_data {
+	struct regulator_common_data common;
+	int32_t current_volt_uv;
+};
+
+static unsigned int regulator_gpio_count_voltages(const struct device *dev)
+{
+	const struct regulator_gpio_config *cfg = dev->config;
+
+	return cfg->states_cnt / 2;
```
Comment (gmarull) [inline drivers/regulator/regulator_gpio.c]: `2U`

File: drivers/regulator/regulator_gpio.c [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+	return 0;
+}
+
+static int regulator_gpio_get_state_voltage(const struct device *dev, int state,
+				       int32_t *volt_uv)
+{
+	const struct regulator_gpio_config *cfg = dev->config;
+	unsigned int state_idx = 0;
+
+	for (; state_idx < cfg->states_cnt / 2; state_idx++) {
```
Comment (gmarull) [inline drivers/regulator/regulator_gpio.c]: `for (unsigned int state_idx = 0U;`

File: drivers/regulator/regulator_gpio.c [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+	if (ret != cfg->states[best_state * 2 + 1]) {
+		ret = gpio_pin_set_dt(&cfg->switch_volt, cfg->states[best_state * 2 + 1]);
+		if (ret < 0) {
+			LOG_ERR("%s: can't set pin state", dev->name);
+			return ret;
+		}
+	}
+
+	data->current_volt_uv = cfg->states[best_state * 2];
+	return 0;
```
Comment (gmarull) [inline drivers/regulator/regulator_gpio.c]: please explain what this code does with some comments. Remember the interface says:

```
The output voltage will be configured to the closest supported output voltage.
```

So no mean between min/max etc.

File: drivers/regulator/regulator_gpio.c [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+	if (idx >= cfg->states_cnt / 2) {
+		LOG_ERR("%s: can't get list voltage for idx %u", dev->name, idx);
+		return -EINVAL;
+	}
+
+	*volt_uv = cfg->states[idx * 2];
+	return 0;
+}
+
+static int regulator_gpio_get_state_voltage(const struct device *dev, int state,
```
Comment (gmarull) [inline drivers/regulator/regulator_gpio.c]: this seems to be an internal helper called at init time, so better move to the top out of API impl.

File: drivers/regulator/regulator_gpio.c [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+
+static int regulator_gpio_init(const struct device *dev)
+{
+	const struct regulator_gpio_config *cfg = dev->config;
+	struct regulator_gpio_data *data = dev->data;
+	int ret;
+
+	regulator_common_data_init(dev);
+
+	if (!device_is_ready(cfg->switch_volt.port)) {
```
Comment (gmarull) [inline drivers/regulator/regulator_gpio.c]: `gpio_is_ready_dt`

File: dts/bindings/regulator/regulator-gpio.yaml [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+
+properties:
+  regulator-name:
+    required: true
+
+  regulator-min-microvolt:
+    required: true
+
+  regulator-max-microvolt:
+    required: true
```
Comment (gmarull) [inline dts/bindings/regulator/regulator-gpio.yaml]: these properties can't be required, they just indicate regulator operating range, which may or may not be set by the user.

File: dts/bindings/regulator/regulator-gpio.yaml [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+
+  switch-voltage-gpios:
```
Comment (gmarull) [inline dts/bindings/regulator/regulator-gpio.yaml]: please keep compatibility with https://www.kernel.org/doc/Documentation/devicetree/bindings/regulator/gpio-regulator.txt, ie use `gpios` and add any missing property, e.g. `startup-delay-us` (except `enable-active-high`, seems to be obsolete).

File: drivers/regulator/regulator_gpio.c [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+		.states = ((const int []) \
+			      DT_PROP(DT_DRV_INST(inst), states)), \
+		.states_cnt = DT_PROP_LEN(DT_DRV_INST(inst), states), \
+	}; \
+	DEVICE_DT_INST_DEFINE(inst, regulator_gpio_init, NULL, &data##inst, \
+			      &config##inst, POST_KERNEL, \
+			      CONFIG_REGULATOR_GPIO_INIT_PRIORITY, \
+			      &regulator_gpio_api);
+
+DT_INST_FOREACH_STATUS_OKAY(REGULATOR_GPIO_DEFINE)
```
Comment (gmarull) [inline drivers/regulator/regulator_gpio.c]: please format this properly, try with ` clang-format`

**Review by gmarull — DISMISSED** [review]

**Comment by xakep-amatop** [comment]
moved regulator gpio driver changes to the next pull https://github.com/zephyrproject-rtos/zephyr/pull/58411

File: include/zephyr/drivers/gpio.h (hunk 2 above)
Comment (xakep-amatop) [reply include/zephyr/drivers/gpio.h]: done

File: drivers/regulator/regulator_gpio.c (hunk 10 above)
Comment (xakep-amatop) [reply drivers/regulator/regulator_gpio.c]: could you please explain what do mean under `properly`?

File: dts/bindings/regulator/regulator-gpio.yaml (hunk 9 above)
Comment (xakep-amatop) [reply dts/bindings/regulator/regulator-gpio.yaml]: Will not copy-pasting everything from Linux kernel be breaking of license rules?

File: drivers/regulator/regulator_gpio.c (hunk 7 above)
Comment (xakep-amatop) [reply drivers/regulator/regulator_gpio.c]: this piece of code as the same as in other regulators with gpios

File: drivers/regulator/regulator_gpio.c (hunk 3 above)
Comment (xakep-amatop) [reply drivers/regulator/regulator_gpio.c]: What of the purpose of this change `2U`?

> suggestion: why don't you store states / 2 to states_cnt? you seem to be dividing by 2 in more than one place
will fix, thanks

File: drivers/regulator/Kconfig.gpio [hunk 11]
```diff
@@ -0,0 +1,20 @@
+# Copyright 2023 EPAM Systems
+# SPDX-License-Identifier: Apache-2.0
+
+config REGULATOR_GPIO
+	bool "GPIO-controlled regulators"
+	default y
+	depends on DT_HAS_REGULATOR_GPIO_ENABLED
+	depends on GPIO
```
[the remark this answers is not in the document]
Comment (xakep-amatop) [reply drivers/regulator/Kconfig.gpio]: fixed

File: drivers/regulator/regulator_gpio.c (hunk 3 above)
Comment (xakep-amatop) [reply drivers/regulator/regulator_gpio.c]: done

File: drivers/regulator/regulator_gpio.c (hunk 5 above)
Comment (xakep-amatop) [reply drivers/regulator/regulator_gpio.c]: fixed

File: drivers/regulator/regulator_gpio.c (hunk 6 above)
Comment (xakep-amatop) [reply drivers/regulator/regulator_gpio.c]: done

File: dts/bindings/regulator/regulator-gpio.yaml (hunk 8 above)
Comment (xakep-amatop) [reply dts/bindings/regulator/regulator-gpio.yaml]: fixed

File: drivers/regulator/regulator_gpio.c (hunk 10 above)
Comment (aasinclair) [reply drivers/regulator/regulator_gpio.c]: https://docs.zephyrproject.org/latest/contribute/guidelines.html#clang-format

**Comment by xakep-amatop** [comment]
Hi, @henrikbrixandersen 
Could you please re-review the pull request?

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]
Looks good. Please update the PR description to match the commit.

**Comment by xakep-amatop** [comment]
> Looks good. Please update the PR description to match the commit.

done

**Review by henrikbrixandersen — APPROVED** [review]

**Review by aasinclair — APPROVED** [review]

**Pull request merged**